src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ArrayBuffer.cs

```diff
@@ -17,46 +18,50 @@ namespace System.Net.Http
     // Commit(byteCount) will extend the ActiveSpan by [byteCount] bytes into the AvailableSpan.
     // Discard(byteCount) will discard [byteCount] bytes as the beginning of the ActiveSpan.
 
-    // TODO: ISSUE 31300: Use ArrayPool to pool buffers.
-
+    [StructLayout(LayoutKind.Auto)]
     internal struct ArrayBuffer : IDisposable
     {
+        private readonly bool _usePool;
         private byte[] _bytes;
         private int _activeStart;
         private int _availableStart;
 
         // Invariants:
         // 0 <= _activeStart <= _availableStart <= bytes.Length
 
-        public ArrayBuffer(int initialSize)
+        public ArrayBuffer(int initialSize, bool usePool = false)
         {
-            _bytes = ArrayPool<byte>.Shared.Rent(initialSize);
-
+            _usePool = usePool;
+            _bytes = usePool ? ArrayPool<byte>.Shared.Rent(initialSize) : new byte[initialSize];
             _activeStart = 0;
             _availableStart = 0;
         }
 
         public void Dispose()
         {
-            byte[] array = _bytes;
+            if (_usePool)
+            {
+                _activeStart = _availableStart = 0;
 
-            _activeStart = _availableStart = 0;
-            _bytes = null;
+                byte[] array = _bytes;
+                _bytes = null;
```

Comment:
This probably doesn't matter in practice, but it seems better to null out _bytes even when we're not using the pool.